source/develop/release-management/features/sla/mom-declarative-language.html.md

```diff
@@ -0,0 +1,276 @@
+---
+title: New policy language for MOM
+category: feature
+authors: akrejcir
+feature_name: New policy language for MOM
+feature_modules: mom
+feature_status: design
+---
+
+# New policy language for MOM
+Currently, [MOM](/documentation/sla/host-mom-policy/)
+policies are written in a lisp-like scripting language
+that can be confusing to use.
+We want to change the model and the language,
+to make it easy for users to write their own policies.
+
+## Owner
+* Owner: Andrej Krejcir
+* Email: akrejcir@redhat.com
+
+## Declarative policy representation
+The rule representation uses YAML format.
+The main idea can be explained on a simple example:
+
+```yaml
+scope: VM
+
+conditions:
+  io_or_net_overutilized:
+    any:
+      - io.read_bytes_per_s  > policy.io_threshold
+      - io.write_bytes_per_s > policy.io_threshold
+      - net.throughput > policy.net_threshold
+
+rules:
+    cpu_limit:
+        output: cpu.max_load
+        min: 10
+        target: 0
+        function:
+            name: linear
+            change: 1
+            time: 1 sec
+
+        when: io_or_net_overutilized
+
+    cpu_release:
+        output: cpu.max_load
+        max: 100
+        function:
+            name: exponential
+            factor: 2
+            time: 30 sec
+
+        when: not io_or_net_overutilized
+```
+
+The first line sets the scope of the rules in the file:
+
+```yaml
+scope: VM
+```
+
+Scope can be:
+
+- `Host` - Rules modify global properties of a host (like KSM). They are executed once per rule evaluation cycle.
+- `VM` - Rules modify properties of VMs. They are executed for all VMs every rule evaluation cycle.
+
+Next is optional definitions of `conditions`.
+They are for convenience, so the same expression is not repeated in multiple rules.
+
+```yaml
+conditions:
+  io_or_net_overutilized:
+    any:
+      - io.read_bytes_per_s  > policy.io_threshold
+      - io.write_bytes_per_s > policy.io_threshold
+      - net.throughput > policy.net_threshold
+```
+
+The condition `io_or_net_overutilized` is defined as the logical `or` of three conditions.
+The objects `io`, `net` and `policy` are VM property objects and
+contain the properties the rules have access to.
+Other property objects can be `cpu`, `memory` or anything else.
+
+The next section defines rules.
+
+```yaml
+rules:
+    cpu_limit:
+        output: cpu.max_load
+        min: 10
+        target: 0
+        function:
```

Comment:
We should probably mention that in this case the same can be achieved using max_absolute_change: 1 / sec.

The function is supposed to decide on what the next value of the output variable is going to be. And we can even say that the default will be immediate jump to `target` (subject to change rate limits).

Reply:
Done. I added a section listing possible functions and mentioning that 'constant' function is default.